CARDS-2611: Support for enforcing specific subject identifier formats

veronikaslc opened the pull request:
Jira issue: https://phenotips.atlassian.net/browse/CARDS-2611

To test the frontend:
- Go to Administration > Subject types
- Create a new Subject Type with:
  - id pattern `^[A-Z]{2}\d{7}$`
  - id pattern hint `2 capital letters followed by 7 digits`
  (OR add these to any existing subject type)
- Go back to the Dashboard
- Start creating a new subject of the type that has the id pattern set

To test the backend: TBD

veronikaslc committed:
CARDS-2611: Support for enforcing specific subject identifier formats

Updated frontend for subject types to support entering and editing of id pattern and its hint

veronikaslc committed:
CARDS-2611: Support for enforcing specific subject identifier formats
Implement a Validator in the data-entry module that prevents the creation of new subjects if an idPattern is specified for their subject type but the provided identifier doesn’t match it

veronikaslc committed:
CARDS-2617: Show error message when a subject id already exists

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -183,10 +193,24 @@ function UnstyledNewSubjectDialog (props) {
           />
+          <div className={classes.newSubjectInput}>
+            <TextField
+              label={!isValid ? "Id does not match the pattern" : "Enter subject identifier"}
```
let's not change the label, it's confusing. Error messages should be in the helper text.

veronikaslc committed:
CARDS-2611: Support for enforcing specific subject identifier formats

Error messages should be in the helper text

marta- commented:
Works great according to specs, but now that I see it in action I'm not a fan of the specs. The interaction is awkward with subject type before the identifier. We need to come up with a better flow.

veronikaslc committed:
CARDS-2611: Support for enforcing specific subject identifier formats
Returned autofocus; validation on type; removed error message about not matching pattern

veronikaslc committed:
CARDS-2611: Support for enforcing specific subject identifier formats
Returned the original order of subject input first and error display

veronikaslc committed:
CARDS-2611: Support for enforcing specific subject identifier formats
Removed on focus error clearance; Error message of duplication of subject removed on type

sashaandjic commented:
@sdumitriu @marta- I am able to modify ID in composum in an invalid format. Strangely I ca add or remove digits, however removing or adding letters was not allowed; the value would go back to the original once I tried to e.g. add another letter.  Please see the screenshot:
![image](https://github.com/user-attachments/assets/9ed7b913-065c-4678-a35a-097a9c8de444)

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/SubjectIdPatternValidator.java:
```diff
@@ -0,0 +1,91 @@
+        try {
+            final String id = subject.getProperty("id").getValue(Type.STRING);
+            // Get the subject's type
+            Node currentParentType = this.subjectTypeUtils.getSubjectType(
```
This variable name is inaccurate.
```suggestion
            Node subjectType = this.subjectTypeUtils.getSubjectType(
```

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/SubjectIdPatternValidator.java:
```diff
@@ -0,0 +1,91 @@
+            }
+        } catch (RepositoryException e) {
+            // Should not happen
+        }
```
Shouldn't we also `catch` a `PatternSyntaxException` ?

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/SubjectIdPatternValidator.java:
```diff
@@ -0,0 +1,91 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.uhndata.cards.forms.internal;
```
Move this file to modules/data-model/subjects/impl/src/main/java/... and update the package:

```suggestion
package io.uhndata.cards.subjects.internal;
```

marta- reviewed (changes requested):
Not sure yet why the backend validator isn't working, but here are some quick comments on its code:

veronikaslc committed:
CARDS-2611: Support for enforcing specific subject identifier formats
Moved validator components to the subject internal module

veronikaslc committed:
CARDS-2611: Support for enforcing specific subject identifier formats
Addressing initial codereview for validators

sashaandjic commented:
I am still able to modify id in wrong format in the composum.

sdumitriu committed:
CARDS-2611: Support for enforcing specific subject identifier formats
Make the validator work

sdumitriu committed:
CARDS-2611: Support for enforcing specific subject identifier formats
Add some logging in case of errors

sdumitriu committed:
CARDS-2611: Support for enforcing specific subject identifier formats
Move the code in the proper module

sdumitriu committed:
CARDS-2611: Support for enforcing specific subject identifier formats
Remove redundant code

sashaandjic reviewed (approved):
tested 
-----
In Composum I can only change the value of identifier if I use the correct pattern. 
Trying to save a value with an invalid pattern, gives an error message shown below:
![image](https://github.com/user-attachments/assets/a48a15b0-11e6-4bb5-a3af-319c7927c30b)

Below is a sample of a modified identifier with a valid pattern, where `AB1234567` was changed to `CC1111111`
![image](https://github.com/user-attachments/assets/76111a1d-bd83-4d46-830d-df6bd792a72c)

sdumitriu commented on modules/data-model/subjects/impl/src/main/java/io/uhndata/cards/subjects/internal/SubjectIdPatternValidator.java:
```diff
@@ -0,0 +1,106 @@
+                    Pattern pattern = Pattern.compile(regexp);
+                    if (!pattern.matcher(id).find()) {
+                        throw new CommitFailedException(CommitFailedException.STATE, 400,
+                            "This subject cannot be created because the id is not validated by subject type pattern");
```
This message needs to be made more user friendly.

marta- replied:
How about  something more explicit, such as "{subjectType} `{id}` cannot be created because its identifier does not match the required format {regexp} ({idPatternHint} /* if it exists */)." ?

veronikaslc replied:
Fixed in [c552736](https://github.com/data-team-uhn/cards/pull/1863/commits/c5527362af6c9ab281f56e230fe57683dd25c669)

I skipped `idPatternHint`, may be the `regexp` is enough for our internal use?

marta- commented on modules/data-model/subjects/impl/src/main/java/io/uhndata/cards/subjects/internal/SubjectIdPatternValidator.java:
```diff
@@ -0,0 +1,106 @@
+            }
+        } catch (PatternSyntaxException e) {
+            throw new CommitFailedException(CommitFailedException.STATE, 400,
+                "This subject cannot be created because the subject type pattern is not valid");
```
"The identifier format requirement for {subjectType} subjects is invalid: {regexp}. No {subjectType} subjects can be created."

veronikaslc replied:
Fixed in [c552736](https://github.com/data-team-uhn/cards/pull/1863/commits/c5527362af6c9ab281f56e230fe57683dd25c669)

veronikaslc committed:
CARDS-2611: Support for enforcing specific subject identifier formats
Improved error messages

sdumitriu reviewed (changes requested):
Bug:
- start in prems mode
- edit the Patient type and set this regexp `^[0-9]{3}$` hint `3 digits`
- edit the Visit type and set this regexp `^[a-z]{3}$` hint `3 letters`
- go to the dashboard, + , New Patient or related entry
- type `abcd` in the name
- select patient -> invalid (OK), select visit -> invalid (OK)
- select patient again
- delete `d` to have `abc` in the name input
- select visit -> invalid (wrong, this should be OK)
- select patient -> valid (wrong, 3 digits expected for patient IDs)
- push Continue -> error (correct, since the serverside validation works and prevents the creation of the subject, but the fact that that I can press Continue is wrong)

It seems that validation is done with the previously selected type, not the new one. It works correctly if I push twice on the right subject type.

This is a shorter sequence of steps but still wrong:
- set up the regexps as before
- start creating a new subject
- type `abc`
- select Visit as the type -> invalid (wrong), click on Visit again -> valid

marta- commented:
> On a side note: shall we restrict adding regexp to types, that already have subjects created? @marta- @sdumitriu

Good catch about the timing of adding the regexp. I wouldn't prevent setting it, but if we want to be really helpful to the admin user I think we should warn them when they try to update the pattern if there might be an issue (ids not matching the pattern): "N subjects of type {subjectType} have identifiers that do not match this pattern. Proceed anyway?"

veronikaslc committed:
CARDS-2611: Support for enforcing specific subject identifier formats
Fixed regex validation

veronikaslc commented:
>  I think we should warn them when they try to update the pattern if there might be an issue 

How do you see this warning in terms of UI?

marta- commented:
> > I think we should warn them when they try to update the pattern if there might be an issue
> 
> How do you see this warning in terms of UI?

When the save button is pressed in the dialog where we enter details about a subject type: 
- if there is no id pattern, proceed with the save
- if there is an id pattern:
  - disable the save button and add a CircularProgress in its center (see the Form save button for how it should look)
  - launch a request to the backend to count all the subjects of that type with IDs that do not match the pattern
  - when the response comes, remove the CircularProgress and re-enable the save button 
  - if the count is zero, proceed with the save
  - if the count is greater than zero, display another dialog layered over the edit dialog
    - Title: **Warning**
    - Text: _{count} subjects of type {subjectType} have identifiers that do not match this pattern. Proceed anyway?_
    - Actions: No (secondary action, button variant: text), Yes (primary action, button variant: contained)
    - when pressing Yes, close the warning dialog, then perform the save of the subject type editing dialog, then close the edit dialog
    - when pressing No, close the warning dialog; keep the other one open for the user to update the pattern

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectTypeDialog.jsx:
```diff
@@ -236,7 +298,7 @@ function SubjectTypeDialog(props) {
               value={idPattern}
               error={isInvalidRegexp}
-              helperText={isInvalidRegexp ? "Invalid regex pattern" : ""}
+              helperText={isInvalidRegexp ? "Invalid regex pattern" : <FormHelperText>{isEdit && currentSubjectType.instanceCount > 0 && "There are already subjects of this type. The parent can no longer be changed"}</FormHelperText>}
```
I don't understand the purpose of this message. Was it supposed to say "pattern" instead of "parent", and is it a leftover from when you suggested to prevent changing the id pattern if some subjects already exist?

We can tell the user that "_Subjects of this type have already been created. Changing the id pattern for this subject type may cause inconsistencies with the existing subject ids._", but not that the pattern can no longer be changed.

One final note: the check of `isEdit` should be outside the `<FormHelperText>` component, so that we don't render an unnecessary component.

veronikaslc replied:
I forgot to remove the message from previous stages of coding, should not be there

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectTypeDialog.jsx:
```diff
@@ -39,6 +40,36 @@ function SubjectTypeDialog(props) {
+    isEdit && currentSubjectType.instanceCount > 0 && fetchData();
+  }, [currentSubjectType]);
+
+  const fetchData = async () => {
+    const urlBase = '/query?query=' + encodeURIComponent('select * from [cards:Subject]');
+    let url = new URL(urlBase, window.location.origin);
+    url.searchParams.set("limit", 1000);
```
Fetching 1000 subjects on page load is not going to give us an accurate view of whether there are subjects that don't match the pattern. There could be more that 1000 subjects overall in the database, and among those that we didn't fetch there could be some that don't match the pattern.

So when we say "2 subjects of type Patient have identifiers that do not match this pattern" we may be underestimating. If in the response for this fetch, the field `totalIsApproximate` is `1`, then we know there's more data we didn't even look at, so we should say "2 OR MORE subjects of type Patient have identifiers that do not match this pattern".

But a worse scenario is when _all_ the subjects whose ids don't match the pattern are among those that we didn't fetch. So we don't detect the problem or warn the user.

2 solutions for this issue:
1. We keep it very **simple** and don't try to detect inconsistencies at all. That means dropping all the code in this commit with the exception of the updated FormHelperText at line 301.

OR

2. We do a **better** job of detecting mismatching subjects by:
  * doing the query just when we are about to update the pattern, ask for subjects of exactly that subject type and pass along the regexp
  * update the query servlet to support matching by regexp
 
  In this case, we know for sure that if there are any mismatching subject Ids we have found them.

I'm ok with doing 1 for now so we can merge this for the CorPath launch, and work on 2 in a separate PR.

veronikaslc replied:
Last commit dropped

sashaandjic reviewed (approved):
tested

sashaandjic commented:
QUESTION: Just to double check that this is intended behavior, as it looks OK to me. 
I created a patient and a visit. Then I changed the pattern for both. 
I created new visit (with pattern), and when I tried to assign it to the existing patient (no pattern) I got this:
![image](https://github.com/user-attachments/assets/e881f456-0ad4-4559-94ae-a06bb6a85f36)
